File: RoutesScala/app/Global.scala
import scala.concurrent.Future

import play.api._
import play.api.mvc._
import play.api.mvc.Results._

import controllers.routes

object Global extends GlobalSettings {

  // Taken from an example by James Roper, member of the Play team
  override def onHandlerNotFound(request: RequestHeader): Future[Result] =
    if (request.path.endsWith("/"))
      Future.successful(MovedPermanently(request.path.take(request.path.length - 1)))
    else
      super.onHandlerNotFound(request)

  // Use this to implement your own router
  //override def onRouteRequest(request: RequestHeader): Option[Handler]

}

File: RoutesJava/app/Global.java
import play.*;
import play.mvc.*;
import static play.mvc.Results.*;

import java.util.*;

import controllers.routes;


public class Global extends GlobalSettings {

  @Override
  public void onStart(Application app) {
    System.out.println(routes.Users.show(1234).url());
  }

  // Taken from an example by James Roper, member of the Play team
  @Override
  public Result onHandlerNotFound(Http.RequestHeader request) {
    if (request.path().endsWith("/"))
      return movedPermanently(request.path().substring(0, request.path().length() - 1));
    else
      return super.onHandlerNotFound(request);
  }
}
